fix(#634): added a generate function to support SQL Numeric field

nisemenov opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->
## Description

- a function for generate random value for SQL NUMERIC type has been added;
- appropriate tests have been added.

I have decided to add a new separate function to work with only SQL NUMERIC type like [this](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL), because [Pydantic Decimal Constraints](https://docs.pydantic.dev/latest/concepts/fields/#decimal-constraints) works in a slightly different way.

Also this function does not allow to use **a negative scale** and **a scale larger than the declared precision**. This can be used only with the SQL standard requirements (the scale must be in the range 0 to precision). And generated values can be either positive or negative.

<!--
Please add in issue numbers this pull request will close, if applicable
Examples: Fixes #4321 or Closes #1234

Ensure you are using a supported keyword to properly link an issue:
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
## Closes
- Closes #634

adhtruong commented:
Based on examples and docs linked, from Pydantic
> class Foo(BaseModel):
>     precise: Decimal = Field(max_digits=5, decimal_places=2)
> 
> 
> foo = Foo(precise=Decimal('123.45')

 and Postgres docs
> So the number 23.5141 has a precision of 6 and a scale of 4. 
I'm not sure the difference.

Based on failing example from issue it looks like an issue in generating the values themselves. Would this fix or similar apply to decimal handling too?

nisemenov commented:
> ...
> 
> how do these differ? Is this for trailing zero handling?
> 
> Based on failing example from issue it looks like an issue in generating the values themselves. Would this fix or similar apply to decimal handling too?

Hi @adhtruong and thanks for reviewing!
Yes, it was my mistake, sorry for that. I didn't fully figure this topic out.

The main thing I have done in last commits:
- added a default value for scale=0 in SQLA module, this field is different from Pydantic (because SQL docs say "NUMERIC(precision) selects a scale of 0.");
- added a constraint check via **max_digits** and **decimal_places** to **get_value_or_none** func.;
- changed **handle_decimal_length** func.

adhtruong commented on polyfactory/value_generators/constrained_numbers.py:
```diff
@@ -317,31 +345,22 @@ def handle_constrained_float(
 
 def validate_max_digits(
     max_digits: int,
-    minimum: Decimal | None,
```
This shouldn't be remove as exported so could be backwards incompatible. Can this be deprecated instead? See `utils.deprecated`

adhtruong commented on polyfactory/value_generators/constrained_numbers.py:
```diff
@@ -317,31 +345,22 @@ def handle_constrained_float(
-    if decimal_places is not None and max_digits <= decimal_places:
-        msg = "max_digits must be greater than decimal places"
+    if decimal_places is not None and max_digits < decimal_places:
+        msg = "max_digits must be greater or equal than decimal places (for the SQL Numeric: the precision must be greater or equal than the scale)"
```
Does this need an explicit mention for SQL Numeric? I think it is preferable fully mapped to parameters so not rely on underlying original type here

adhtruong commented on polyfactory/factories/sqlalchemy_factory.py:
```diff
@@ -175,6 +175,8 @@ def _get_type_from_type_engine(cls, type_engine: TypeEngine) -> type:
             for sqlalchemy_field, constraint_field in constraint_fields.items():
                 if (value := getattr(type_engine, sqlalchemy_field, None)) is not None:
                     constraints[constraint_field] = value  # type: ignore[literal-required]
+                if sqlalchemy_field == "precision":
```
Based on PSQL, I'm not sure this works for that case as will always set `decimal_places` instead of being unconstrained.  I think this is safe to omit and require user to specify if require this behaviour.

Alternatively, I think may be able to push to `get_constraint_fields` then would allow this behaviour to be located with the column type?

nisemenov replied:
Ok, but I mean this:

> Alternatively: NUMERIC(precision) selects a scale of 0.

Could you explain in more detail, please?

adhtruong replied:
This currently is basing this on `sqlalchemy_field` only rather than targeting `Numeric` type specifically.  

If a user overrides, `get_constraint_fields` with something that specifies `precision` then `scale` will be set. This behaviour should be co-located with the type if possible. Maybe by adapting `get_constraints_fields` to return a callable from instance of TypeEngine to return constraints? 

I think it's think to omit this default and require user to specify if they want this behaviour. This reduces the amount of custom logic per factory and SQL-specific knowledge in polyfactory itself

adhtruong commented on polyfactory/value_generators/constrained_numbers.py:
```diff
@@ -357,25 +376,18 @@ def handle_decimal_length(
-
-    decimals = decimals[:max_decimals]
-    return Decimal(sign + whole_numbers + "." + decimals[:decimal_places])
+    with localcontext() as ctx:
```
Nice implemention!
